filters/diag/weakrand.go

```diff
@@ -0,0 +1,46 @@
+package diag
+
+import (
+	"math/rand"
+	"sync"
+)
+
+type reader struct {
+	c  []byte
+	i  int
+	mx *sync.Mutex
+}
+
+var randomChars = []byte("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789")
+
+func weakRandom() *reader {
+	c := make([]byte, len(randomChars))
+	copy(c, randomChars)
+	rand.Shuffle(len(c), func(i, j int) {
+		c[i], c[j] = c[j], c[i]
+	})
+
+	return &reader{c: c, mx: &sync.Mutex{}}
+}
+
+func (r *reader) Read(p []byte) (int, error) {
+	r.mx.Lock()
+	i := r.i
+	r.mx.Unlock()
+
+	var n int
+	for len(p) > 0 {
+		l := copy(p, r.c[i:])
+		p = p[l:]
+		n += l
+		i += l
+		if i >= len(r.c) {
+			i = 0
+		}
+	}
+
+	r.mx.Lock()
```

Comment:
I think we know that returned length is always `len(p)` and also do not need second mutex lock:
```diff
diff --git a/filters/diag/weakrand.go b/filters/diag/weakrand.go
index 4ad7185..4c92ff5 100644
--- a/filters/diag/weakrand.go
+++ b/filters/diag/weakrand.go
@@ -24,23 +24,20 @@ func weakRandom() *reader {
 }
 
 func (r *reader) Read(p []byte) (int, error) {
+       n := len(p)
+
        r.mx.Lock()
        i := r.i
+       r.i = (r.i + n) % len(r.c)
        r.mx.Unlock()
 
-       var n int
        for len(p) > 0 {
                l := copy(p, r.c[i:])
                p = p[l:]
-               n += l
                i += l
-               if i >= len(r.c) {
+               if i == len(r.c) {
                        i = 0
                }
        }
-
-       r.mx.Lock()
-       r.i = i
-       r.mx.Unlock()
        return n, nil
 }
```

Reply:
nice! Thanks a lot!